46_revocation_severity_and_context.md

```diff
@@ -0,0 +1,301 @@
+validating further. The state of the agent after that should be
+`QUOTE_FAILED_IRRECOVERABLE`.
+
+The highest severity is 0 and the higher the number the number the lower the
+severity. This makes it easy to add less sever levels without the need to
+increase the level of the highest severity.
+
+A new table `revocation_events` is introduced to save the already send
+notifications. It contains the columns:
+  * `agent_id`: The id of agent where the event was created.
+  * `event_id`: The event ID string
+  * `severity_level`: The severity of the event.
+  * `context`: String or JSON object that contains more information about that event.
+
+In the `process_agent` function gets a new argument that can contain a failure
+object. If the status is `QUOTE_FAILED` the events from failure object are
+checked against the database and if there are occurring the first time a message
+is send and they are added to the database. Otherwise the event is ignored.
+
+In the verifier API for the agent a new field is exposed that contains the list
+of the revocation events, with their respected level and context.
+
+To the send revocation message two new field are added: `severity_level` and
+`context`. Such that the other agents can act accordingly.
+
+(To reduce bandwidth all the generated messages could also be bundled into one,
+but that would introduce more complexity of parsing the messages for the
+recipient.)
+
+If a agent gets removed from the verifier also all entries in
+`revocation_events` for the agent will be removed. (An option to clear the old
+events from a agent could be introduced, but a reset can already implemented by
+removing and then adding the agent to the verifier)
+
+### Events
+Events are tagged with an event id. Which has the following schema
+`component.[sub_component].event`.
+
+* `component`: Name of the part of Keylime where the event was generated.
+  * Current would be: `quote_validation`, `pcr_validation`, `measured_boot`,
+    `ima` and `internal`.
+* `sub_component`: If it is useful the separate a component into other sub
+  components. One example would be IMA checks.
+* `event`: The actual event itself.
+An example event id for a failing static PCR check would be `pcr_validation.pcr10`.
+
+In most cases the event ids static but for dynamic policies the `sub_component`
+and `event` can be generated automatically.
+
+The motivation behind that schema is to allow Keylime to adjust granularity of
+the events where needed and give the user an easy way to specify a severity
+level to one whole subset of Keylime.
+
+The severity level of an event is either 0 or if it matches one of the user
+rules the according to the configuration.
+
+Events can a specified context. Those can be either a static string or an JSON
+object that contains more information event. This is optional and it is assumed
+that two events from the same agent with the same event id have the same
+severity and are the same event.
```

Comment:
I don't think we can assume they are for the same event right? It's possible to have the same failure one hour, recover and have a similar failing again the next hour right? Or am I missing something.

Reply:
I think my wording for recoverable vs irrecoverable events is not precise enough. What I meant with recoverable was that Keylime can still operate and do (parts of) remote attestation. Is would case for an IMA failure but not if the qoute cannot be validated (and retries failed).

In my opinion you cannot really recover from a revocation event because in most cases that just means that a window for attack was open and the system might be compromised.
If it was determined that window was not used maybe the events can deleted manually or reset the agent by removing and adding it to the verifier again.

The main idea behind that is, that the receiver of the revocation events does not get a message every time when the failed agent is polled if it was already notified. 
Another idea was that our decision is mainly based on the id and context is more for deeper analysis if needed. 
 
But I think if the context part is heavily used i could make sense to make it not optional and consider two events with the same id but different context as two separate events.